sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaThriftTable.java

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.kafka;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.List;
+import org.apache.beam.sdk.io.thrift.ThriftCoder;
+import org.apache.beam.sdk.io.thrift.ThriftSchema;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.SchemaProvider;
+import org.apache.beam.sdk.schemas.transforms.Convert;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.Values;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TFieldIdEnum;
+import org.apache.thrift.protocol.TProtocolFactory;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+public class BeamKafkaThriftTable<FieldT extends TFieldIdEnum, T extends TBase<T, FieldT>>
+    extends BeamKafkaTable {
+  private final ThriftCoder<T> thriftCoder;
+  private final TypeDescriptor<T> typeDescriptor;
+
+  public BeamKafkaThriftTable(
+      @NonNull Schema requiredSchema,
+      @NonNull String bootstrapServers,
+      @NonNull List<String> topics,
+      @NonNull Class<T> thriftClass,
+      @NonNull TProtocolFactory protocolFactory) {
+    super(thriftSchema(thriftClass, requiredSchema), bootstrapServers, topics);
+    typeDescriptor = TypeDescriptor.of(thriftClass);
+    thriftCoder = ThriftCoder.of(thriftClass, protocolFactory);
+  }
+
+  private static Schema thriftSchema(
+      @NonNull Class<?> thriftClass, @NonNull Schema requiredSchema) {
+    @SuppressWarnings("nullness")
+    final @NonNull Schema thriftSchema =
+        ThriftSchema.provider().schemaFor(TypeDescriptor.of(thriftClass));
+    if (!requiredSchema.equivalent(thriftSchema)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Given message schema: '%s'%n"
+                  + "does not match schema inferred from thrift class.%n"
+                  + "Thrift class: '%s'%n"
+                  + "Inferred schema: '%s'",
+              requiredSchema, thriftClass.getName(), thriftSchema));
+    }
+    return thriftSchema;
+  }
+
+  @Override
+  protected PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>> getPTransformForInput() {
+    final @NonNull SchemaProvider schemaProvider = ThriftSchema.provider();
+    return new PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>>() {
+      @Override
+      @SuppressWarnings("nullness")
+      public PCollection<Row> expand(PCollection<KV<byte[], byte[]>> input) {
+        return input
+            .apply(Values.create())
+            .apply(MapElements.into(typeDescriptor).via(BeamKafkaThriftTable.this::decode))
```

Comment:
Wouldn't this::decode be sufficient?

Reply:
Just one more comment.
It would be good to have some function or ParDo declared like it is done for protobuf so we don't have to repeat methods like encode and decode together with ThriftCoder in every table provider that supports thrift.
We could then just import a function returning SimpleFunction/ParDo and use it like `MapElements.via(thriftBytesToBeamRow(thriftClass, protocolFactory))`